src/plugins/ui_actions/public/service/ui_actions_service.ts

```diff
@@ -0,0 +1,193 @@
+    this.triggers = triggers;
+    this.actions = actions;
+    this.triggerToActions = triggerToActions;
+  }
+
+  public readonly registerTrigger = (trigger: Trigger) => {
+    if (this.triggers.has(trigger.id)) {
+      throw new Error(`Trigger [trigger.id = ${trigger.id}] already registered.`);
+    }
+
+    this.triggers.set(trigger.id, trigger);
+    this.triggerToActions.set(trigger.id, []);
+  };
+
+  public readonly getTrigger = (id: string) => {
+    const trigger = this.triggers.get(id);
+
+    if (!trigger) {
+      throw new Error(`Trigger [triggerId = ${id}] does not exist.`);
+    }
+
+    return trigger;
+  };
+
+  public readonly registerAction = (action: Action) => {
+    if (this.actions.has(action.id)) {
+      throw new Error(`Action [action.id = ${action.id}] already registered.`);
+    }
+
+    this.actions.set(action.id, action);
+  };
+
+  public readonly attachAction = (triggerId: string, actionId: string): void => {
+    const trigger = this.triggers.get(triggerId);
+
+    if (!trigger) {
+      throw new Error(
+        `No trigger [triggerId = ${triggerId}] exists, for attaching action [actionId = ${actionId}].`
+      );
+    }
+
+    const actionIds = this.triggerToActions.get(triggerId);
+
+    if (!actionIds!.find(id => id === actionId)) {
+      this.triggerToActions.set(triggerId, [...actionIds!, actionId]);
+    }
+  };
+
+  public readonly detachAction = (triggerId: string, actionId: string) => {
+    const trigger = this.triggers.get(triggerId);
+
+    if (!trigger) {
+      throw new Error(
+        `No trigger [triggerId = ${triggerId}] exists, for detaching action [actionId = ${actionId}].`
+      );
+    }
+
+    const actionIds = this.triggerToActions.get(triggerId);
+
+    this.triggerToActions.set(
+      triggerId,
+      actionIds!.filter(id => id !== actionId)
+    );
+  };
+
+  public readonly getTriggerActions = (triggerId: string) => {
+    // This line checks if trigger exists, otherwise throws.
+    this.getTrigger!(triggerId);
+
+    const actionIds = this.triggerToActions.get(triggerId);
+    const actions = actionIds!
+      .map(actionId => this.actions.get(actionId))
+      .filter(Boolean) as Action[];
+
+    return actions;
+  };
+
+  public readonly getTriggerCompatibleActions = async <C>(triggerId: string, context: C) => {
+    const actions = this.getTriggerActions!(triggerId);
+    const isCompatibles = await Promise.all(actions.map(action => action.isCompatible(context)));
+    return actions.reduce<Action[]>(
+      (acc, action, i) => (isCompatibles[i] ? [...acc, action] : acc),
+      []
+    );
+  };
+
+  private async executeSingleAction<A extends {} = {}>(action: Action<A>, actionContext: A) {
+    const href = action.getHref && action.getHref(actionContext);
+
+    if (href) {
+      window.location.href = href;
+      return;
+    }
+
+    await action.execute(actionContext);
+  }
+
+  private async executeMultipleActions<C>(actions: Action[], actionContext: C) {
```

Comment:
What do you think about changing the name of this?  I was thinking in the future it might be nice for the action itself to specify whether it should be "auto executed" or have the user choose?  Thinking like in the context of an on hover action - if you wanted to both show a tooltip pop up _and_  show a global threshold bar on all the other embeddables, these are two actions you'd want auto executed and not ask the user to choose between them.

But, we don't actually have that use case right now.  Up to you.

Reply:
This is an internal `private` helper function, I'm happy giving it any name, have 0 preferences here.